k2/python/k2/mwer_loss.py

```diff
@@ -0,0 +1,436 @@
+        """
+        scores_shape = self.fsa.arcs.shape().remove_axis(1)
+        # scores_shape has axes [path][arc]
+
+        ragged_lm_scores = k2.RaggedTensor(
+            scores_shape, self.fsa.lm_scores.contiguous()
+        )
+
+        tot_scores = ragged_lm_scores.sum()
+
+        return k2.RaggedTensor(self.shape, tot_scores)
+
+    def tot_scores(self) -> k2.RaggedTensor:
+        """Get total scores of FSAs in this Nbest.
+
+        Note:
+          Since FSAs in Nbest are just linear FSAs, log-semiring
+          and tropical semiring produce the same total scores.
+
+        Returns:
+          Return a ragged tensor with two axes [utt][path_scores].
+          Its dtype is torch.float64.
+        """
+        scores_shape = self.fsa.arcs.shape().remove_axis(1)
+        # scores_shape has axes [path][arc]
+
+        ragged_scores = k2.RaggedTensor(
+            scores_shape, self.fsa.scores.contiguous()
+        )
+
+        tot_scores = ragged_scores.sum()
+
+        return k2.RaggedTensor(self.shape, tot_scores)
+
+    def build_levenshtein_graphs(self) -> k2.Fsa:
+        """Return an FsaVec with axes [utt][state][arc]."""
+        word_ids = get_texts(self.fsa, return_ragged=True)
+        return k2.levenshtein_graph(word_ids)
+
+
+class MWERLoss(nn.Module):
+    # Minimus Word Error Rate.
+    # Equation 2 of https://arxiv.org/pdf/2106.02302.pdf
+    def __init__(self,):
+        super().__init__()
+
+    def forward(self, lattice, ref_texts, nbest_scale, num_paths):
+        use_double_scores = True
+
+        nbest = Nbest.from_lattice(
+            lattice=lattice,
+            num_paths=num_paths,
+            use_double_scores=use_double_scores,
+            nbest_scale=nbest_scale,
+        )
+
+        device = lattice.scores.device
+        hyps = nbest.build_levenshtein_graphs()
+        refs = k2.levenshtein_graph(ref_texts, device=hyps.device)
+        levenshtein_alignment = k2.levenshtein_alignment(
+            refs=refs,
+            hyps=hyps,
+            hyp_to_ref_map=nbest.shape.row_ids(1),
+            sorted_match_ref=True,
+        )
+        tot_scores = levenshtein_alignment.get_tot_scores(
+            use_double_scores=False, log_semiring=False
+        )
+        # Each path has a corresponding wers.
+        tot_num_paths = nbest.kept_path.tot_size(0)
+        tot_num_arcs = nbest.kept_path.tot_size(1)
+        # So its shape is [tot_num_paths].
+        wers = -tot_scores
+
+        nbest_logp = nbest.fsa.scores
+
+
+        # Group each log_prob into [path][arc]
+        ragged_nbest_logp = k2.RaggedTensor(nbest.kept_path.shape, nbest_logp)
+        # Get the probalitity of each path, in log format, with shape [path].
+        path_logp = ragged_nbest_logp.sum()
+
+        # Group path_logp into [stream][path] to compute denominator of each stream.
+        # A stream here means an input wav.
+        ragged_path_prob = k2.RaggedTensor(nbest.shape, path_logp.exp())
```

Comment:
You could perhaps write a softmax function accepting RaggedTensor to replace these few lines, and  call that instead, it will be clearer.

Reply:
A function `k2.ragged_softmax` is added to k2/ops.py.